fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -2444,13 +2447,32 @@ private Bundle createSearchBundle(List<Resource> resources, FHIRSearchContext se
                                             .id(UUID.randomUUID().toString())
                                             .total(totalCount);
 
+        // Calculate how many resources are 'match' mode
+        int pageSize = searchContext.getPageSize();
+        int offset = (searchContext.getPageNumber() - 1) * pageSize;
+        int matchResourceCount = pageSize;
+        if (totalCount.getValue() < offset + pageSize) {
+            matchResourceCount = totalCount.getValue() - offset;
+        }
+
+        // Check if too many included resources
+        if (resources.size() > matchResourceCount + SearchConstants.MAX_PAGE_SIZE) {
+            String msg = "Number of returned 'include' resources exceeds allowable limit of " + SearchConstants.MAX_PAGE_SIZE;
+            throw buildRestException(msg, IssueType.BUSINESS_RULE, IssueSeverity.ERROR);
+        }
+
         for (Resource resource : resources) {
             if (resource.getId() == null) {
                 throw new IllegalStateException("Returned resources must have an id.");
             }
-            Bundle.Entry entry = Bundle.Entry.builder().fullUrl(Uri.of(getRequestBaseUri(type) + "/"
-                    + resource.getClass().getSimpleName() + "/"
-                    + resource.getId())).resource(resource).build();
+            Bundle.Entry entry = Bundle.Entry.builder()
+                    .fullUrl(Uri.of(getRequestBaseUri(type) + "/" + resource.getClass().getSimpleName() + "/" + resource.getId()))
+                    .resource(resource)
+                    .search(Search.builder()
+                        .mode(matchResourceCount-- > 0 ? SearchEntryMode.MATCH : SearchEntryMode.INCLUDE)
```

Comment:
why are we decrementing `--` matchResourceCount (is this an off by one problem?)  If that's the case, shouldn't it be prefixed `--`?

Reply:
also since matchResourceCount is calculated one time, why not create a local variable searchEntryMode instead of looping the calculation?

Reply:
Regardless of how we do it, we have to do a compare of the iteration we're on inside the loop to 'matchResourceCount' in order to set the mode properly. I'm simply decrementing 'matchResourceCount' each time through the loop, and when we get to 0, we know we have to switch the mode from 'match' to 'include'. I don't think there's a more efficient way to do it, but if you think the code is more readable using a local variable for the mode I can do that.

Reply:
Got it.  Do you mind adding a comment on the logic to the code? I see this as a hidden bit of logic, and it'd be good to surface it in a comment.

Reply:
Comment added.